### ISSUE TITLE
Touch keyboard is not displayed properly on Windows

### ISSUE THREAD

**Issue by mgnslndh** [AvaloniaUI/Avalonia#10136]
Touching a `TextBox` does not bring up the keyboard. Touching "native" Windows text input like Notepad or similar will bring up the keyboard.

*Expected behavior*
When I touch a textbox on a touch screen I want this keyboard to be displayed:

![image](https://user-images.githubusercontent.com/1139118/212348315-de1680a1-98a3-4082-b48f-64beff2626d4.png)

- Avalonia
  - Version 11.0.preview4
- Windows
  - Edition: Windows 10 Pro
  - Version: 21H2
  - OS Build: 19044.2364

_Originally posted by @mgnslndh in https://github.com/AvaloniaUI/Avalonia/discussions/9974

**Comment by maxkatz6** [comment]
@mgnslndh that's also something what we already support for mobile. 
I.e.:
```
<TextBox Watermark="Pin" PasswordChar="*" TextInputOptions.ContentType="Digits" />
```
Same API should be used for the windows.

**Comment by mgnslndh** [comment]
@maxkatz6 that sounds awesome. I have no idea what is broken but I'm willing to help out with testing or fixing the issue if someone can troubleshoot it to find the root cause.

**Comment by maxkatz6** [comment]
@mgnslndh tbh I am not an expert of IME integration, especially on windows. Maybe @Gillibald @yatli @Mikolaytis know better.

"imm32" is used for the IME support, and specifically ImmSetCandidateWindow is invoked to ask OS to show language specific IME window. It works with languages complex text input like Chinese and mouse input, but for some reason it doesn't show the keyboard on touch input for any other language.
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Input/Imm32InputMethod.cs#L244

**Comment by Sorien** [comment]
maybe this could help https://stackoverflow.com/a/40921638 and for changing layout setinputscope should be working https://learn.microsoft.com/en-us/windows/win32/api/inputscope/nf-inputscope-setinputscope

**Comment by mgnslndh** [comment]
@maxkatz6 @Sorien 
Ok, so there are multiple ideas of how to control the touch keyboard on Windows. But there are some fundamentals that I would like to understand:

1. Is touch input for languages such as English considered complex text input and handled by Imm32? Or Should it belong in a separate implementation of `ITextInputMethodImpl` ?
2. Do we know if Imm32InputMethod works as intended for complex text input such as Chinese or is it broken, just like for ordinary touch with English language? Is it possible to test complex text input on Windows while still using English? 
3. Is it the same Windows component (tabtip.exe) that will be displayed when `Imm32InputMethod` is used with complex language input such as Chinese, or will some other Windows component be displayed?
4. If the Windows component displayed is not TabTip.exe, which is used for touch input, how would we determine if one or the other should be displayed?

I guess the questions above are kind of flavors of the same one question :)

**Comment by tobyfirth** [comment]
Hi,

I have been doing a little bit of research on this. From my reading of [Respond to the presence of the touch keyboard](https://learn.microsoft.com/en-us/windows/apps/design/input/respond-to-the-presence-of-the-touch-keyboard) it seems to imply that the touch keyboard appearance is controlled by UI Automation. 
The AutomationPeer classes appear to be present in Avalonia so that might not be the case then.

For the input scope to show different keyboard layouts it looks like the interfaces from [Text Services Framework](https://learn.microsoft.com/en-us/windows/win32/api/_tsf/) would need to be implemented. Using the SetInputScope function is a possibility but the documentation suggests that applications should not be using this directly and implement the interfaces instead ([ITfInputScope interface](https://learn.microsoft.com/en-us/windows/win32/api/inputscope/nn-inputscope-itfinputscope)).

**Comment by mozesa** [comment]
Hello! I would like to inquiry if there is a chance to get working  native OSK on Windows?
Unfortunately, this fix was closed months ago and still no progress 😞

**Comment by timunie** [comment]
@mozesa if you know any, your contribution is welcome. Otherwise you will have to wait until we have one.

**Comment by mozesa** [comment]
@timunie I am starting to use this: [https://gist.github.com/mozesa/7f203bb926dab37eb41d783933cb48d2](url).
Unfortunately, it is not the proper way for Avalonia.

**Comment by tobyfirth** [comment]
@mozesa I have done something similar to you to get the touch keyboard to show for my project.

I found this [library ](https://github.com/maximcus/WPFTabTip/tree/master) for WPF which has some helpful ideas for how to develop a solution.

I also found that if polling for the keyboard window sizes to check if already open/closed there were some small differences between windows 10 and 11 in what was reported.

**Comment by mozesa** [comment]
@tobyfirth I started also with that library, unfortunately that approach requires elevated rights (i.e. admin).

**Comment by mozesa** [comment]
I leave it here.
There is eight-year [bug description](https://bugzilla.mozilla.org/show_bug.cgi?id=1200799).
The below code works, I mean, when there's touch capable monitor attached to the PC, and I click into the `TextBox`, the OSK shows up. But unfortunately, these methods are not reliably hiding - showing the OSK. I tried to specify not the `Window` but the `Context` but it would like to use `UIContext`. All in all, I am continuing...  

```csharp
using Windows.UI.ViewManagement;
using Adapter.Desktop.Views;
using Avalonia;
using Avalonia.Controls;
using Avalonia.Input;
using Avalonia.Interactivity;

namespace Adapter.Desktop.Pages;

public partial class ItemsPage : UserControl
{
    private IntPtr _handle;
    private InputPane _inputPane;

    public ItemsPage()
    {
        InitializeComponent();
    }

    protected override void OnAttachedToVisualTree(VisualTreeAttachmentEventArgs e)
    {
        base.OnAttachedToVisualTree(e);

        _handle = ((MainWindow)VisualRoot!).TryGetPlatformHandle()!.Handle;
        _inputPane = InputPaneInterop.GetForWindow(_handle);
    }

    private void InputElement_OnGotFocus(object? sender, GotFocusEventArgs e)
    {
        Console.WriteLine($"Show: {_inputPane.TryShow()}");
    }

    private void InputElement_OnLostFocus(object? sender, RoutedEventArgs e)
    {
        Console.WriteLine($"Hide: {_inputPane.TryHide()}");
    }
}
```

```csproj
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <TargetFramework>net7.0-windows10.0.19041.0</TargetFramework>
```

**Comment by workgroupengineering** [comment]
see [this](https://github.com/microsoft/WPF-Samples/tree/main/Input%20and%20Commands/TouchKeyboard/TouchKeyboardNotifier)

**Comment by tobyfirth** [comment]
@mozesa In terms of the library that I referenced it was the Observable sequence used in it that I found helpful. Actually showing/hiding the keyboard I use the ITipInvocation interface like you do in your gist.

I could never get IFrameworkInputPane interface to work for me.

I used AddClassHandler<TextBox> on GotFocusEvent, PointerPressedEvent, LostFocusEvent and KeyDownEvent to get the events which I fed into an observable sequence.

**Comment by Mr4g** [comment]
Hello,
if i use  TextInputOptions.ContentType="Digits" my numeric virtual keyboard doesn't open.

![image](https://github.com/AvaloniaUI/Avalonia/assets/133530721/0756aafe-e817-48f2-a177-857665a52941)

**Comment by tobyfirth** [comment]
I had another look at this and I think I have found what part of the UI Automation the touch keyboard responds to.
According to this Microsoft example repo [https://github.com/microsoft/Windows-universal-samples/tree/93bdfb92b3da76f2e49c959807fc5643bf0940c9/Samples/TouchKeyboard](https://github.com/microsoft/Windows-universal-samples/tree/93bdfb92b3da76f2e49c959807fc5643bf0940c9/Samples/TouchKeyboard)

> On the PC, you can request that the touch keyboard display for a custom control by implementing the TextPattern provider interface ([ITextProvider](http://msdn.microsoft.com/library/windows/apps/br242627)) and the ValuePattern provider interface ([IValueProvider](http://msdn.microsoft.com/library/windows/apps/br242663)). Not supported on Phone.

This line is not in the latest version of the repo though.

I also found this post [https://blog.tombam.net/implementing-textbox-with-on-screen-touch-keyboard-part-2/](https://blog.tombam.net/implementing-textbox-with-on-screen-touch-keyboard-part-2/) which implements the ITextProvider and IValueProvider interfaces in WinForms. I have tested the code from the post and it does trigger the touch keyboard when the interfaces are used and the keyboard is not triggered when they are not so it is definitely the presence of these interfaces that controls the triggering of the touch keyboard.

It looks like IValueProvider is already implemented for Avalonia which leaves the ITextProvider (and ITextRangeProvider which is uses) to be implemented. Both of those interfaces are already defined in the Interop directory for Win32.

**Comment by timunie** [comment]
@tobyfirth if you have an idea how to implement that, consider to add a draft-PR for feedback and testing.

**Comment by tobyfirth** [comment]
@timunie I would like to do that. Not quite sure when I will have the time though. Hopefully in the next few weeks.

**Comment by Hackmodford** [comment]
@tobyfirth Did you ever get time to implement this?

**Comment by tobyfirth** [comment]
@Hackmodford Unfortunately not yet. I think I might be able to start it this week or next though.

**Comment by Hackmodford** [comment]
@tobyfirth Is the idea just that you'd need to add the `ITextProvider` to something like [this](https://github.com/AvaloniaUI/Avalonia/blob/0ae36f98ba4deefce188e1bd8026f523d2570e20/src/Avalonia.Controls/Automation/Peers/TextBoxAutomationPeer.cs)?

**Comment by tobyfirth** [comment]
@Hackmodford Yes. And also the `ITextRangeProvider`. That is a little trickier as it has to deal with selecting a range of text and there are quite a few of functions to write.

I have now got a basic implementation, basically a copy of [https://blog.tombam.net/implementing-textbox-with-on-screen-touch-keyboard-part-2/](https://blog.tombam.net/implementing-textbox-with-on-screen-touch-keyboard-part-2/). I haven't actually tested it with a touch screen as the computer that I can write the code on does not have one. It looks alright when inspecting it with Accessibility Insights for Windows, at least it doesn't cause and crashes or exceptions. Hopefully tomorrow I can test the sandbox on a touch screen.

The main issue is that the code does not properly support selecting text ranges which will need some more work.

**Comment by federicocodo** [comment]
@tobyfirth Can you share the code? I too would need the OSK for avalonia for a project, I also have a Windows touch screen so I can check if it works and maybe help with the code

**Comment by tobyfirth** [comment]
I have push my not working version to my [github account](https://github.com/tobyfirth/Avalonia/tree/WindowsAutomationTextBox).
Maybe someone else might be able to see something that I am missing.

Another possibility is that there is something that [https://blog.tombam.net/implementing-textbox-with-on-screen-touch-keyboard-part-2/](https://blog.tombam.net/implementing-textbox-with-on-screen-touch-keyboard-part-2/) example has that has been provided by default by winforms that the current avalonia implementation is missing.

**Comment by Laurnz** [comment]
I test sometimes test touchscreen features on non touchscreen Windows device [like here](https://stackoverflow.com/a/40274661). You have to install Visual Studio 2019, it does not come with 2022. Also, do not forget to configure the touchscreen to show even when a keyboard is attached in the Windows settings.  
<br/>

> @tobyfirth I have tested the code from the post and it does trigger the touch keyboard when the interfaces are used and the keyboard is not triggered when they are not

For me, the touch keyboard gets activated whether the interfaces are there or not in https://blog.tombam.net/implementing-textbox-with-on-screen-touch-keyboard-part-2/. Maybe I am doing something wrong, because the automation provider never even got instantiated for the example or the Avalonia implementation.

**Comment by tobyfirth** [comment]
That is odd. I will have a look at that program.

**Comment by Laurnz** [comment]
It seems like Microsoft has the Text Services Framework that it is being used for all on-screen keyboard related tasks. It apparently internally handles showing/hiding the OSK. The TSF is explained [by the MSEdge team here](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/TSF1/explainer.md).

[Here](https://www.dotnetframework.org/default.aspx/DotNET/DotNET/8@0/untmp/WIN_WINDOWS/lh_tools_devdiv_wpf/Windows/wcp/Core/System/Windows/Input/TextServicesContext@cs/2/TextServicesContext@cs) they are using it for WPF.

Maybe it would be worthwhile to also implement this for Avalonia.

**Comment by tobyfirth** [comment]
Using the windows simulator I didn't get the touch keyboard appearing when using my test app for avalonia but it did for the winforms app although I didn't try it with the added interfaces disabled.

Text Services Framework will be needed for things like showing the different keyboard layouts. When I originally looked at it seemed like it would be a lot more work to implement than adding `ITextProvider` and `ITextRangeProvider`.

And it doesn't look like Text Services Framework is needed as the core interface for that `ITextStoreACP` is not implemented in the .Net version of WinForms so was unlikely to have been in the framework version but the touch screen keyboard can still be made to appear by adding the UI Automation interfaces.

**Comment by maxkatz6** [comment]
@Laurnz yes, implementing TSF is definitely something we want to have at some point. But it's rather a substantial task, and not planned for next releases.

**Comment by Arthur-Camatti** [comment]
@tobyfirth Have you found any luck in making your version work?

Also, I'm too able to use the OSK on the example from Tombam's blog without the interfaces, even after replacing the custom TextBox with the default one. Seems like WinForms got an update on this since this solution.

**Comment by tobyfirth** [comment]
@Arthur-Camatti, I have also now found that the interfaces do not actually do anything anymore. Maybe at some point in the past. TSF looks like the correct way to get windows to recognise a text input and to trigger the touch keyboard if appropriate.

However, Avalonia does not currently support TSF so another way is needed. Since the release of 11.1 with the `IInputPane` interface this is a lot simpler. This [gist](https://gist.github.com/tobyfirth/65c5372be2e659141c1c4b7d99e3e268) is the current solution that I have created. I have only tested it as a proof of concept locally so there are probably edge cases that I have not encountered yet. It is also limited by the fact that I only want to use it for a fullscreen app so the code for moving the UI out of the way assumes that the window is fullscreen. If using with just a maximised window the the code for shifting the window is probably not what you want and you will need to come up with an appropriate alternative. 

I hope that the code is still useful for you with these limitations.

**Comment by Ammar-Abush** [comment]
Any updates on this issue?

**Comment by lindexi** [comment]
Reference: https://github.com/dotnet/wpf/issues/10427

**Comment by giard-alexandre** [comment]
This will probably go into the garbage once Avalonia has proper support for Windows tablets but in the meantime, I created an easy-to-consume (and hopefully easy-to-use) package for this 😄 
[TabTip.Avalonia](https://github.com/giard-alexandre/TabTip.Avalonia)

**Comment by Arthur-Camatti** [comment]
I have a question: now that an On-Screen Keyboard is available on [Avalonia Accelerate's paid Enterprise plan](https://avaloniaui.net/accelerate/on-screen-keyboard), what happens to this issue? Windows has it's own OSK but Avalonia fails to integrate it, and from the changelogs since this issue was created, nothing has changed. As I see, support for this shouldn't be made via a "monkey patch" like the one on Toby's gist, much less be gated behind a paywall.

**Comment by tobyfirth** [comment]
My understanding of the On-Screen Keyboard included in the enterprise plan is that it is targeting very different use cases. It is either for systems that are so minimal that they don't come with a system OSK (e.g. very small Linux installs which don't even ship with a desktop environment so you have to supply your own OSK) and very locked down systems (e.g. the enterprise plan mentions kiosk systems so something which any member of the public can interact with) where using the system OSK would allow the user to access things they shouldn't like being able to open the start menu on Windows. Avalonia not triggering it by default on systems which lack a physical keyboard, would make it slightly easy to implement for the second use case.

The main reason is that for Windows it looks like you have to implement quite a complicated API for the system to automatically show the keyboard for you. And the use cases the enterprise plan is targeting are likely to use Linux just as much as Windows and when you are using Windows you don't want to use the system OSK for security reasons so implementing the Windows APIs does not solve the problem they are trying to solve with that keyboard.
